u2f-ref-code/java/src/com/google/u2f/server/impl/U2FServerReferenceImpl.java

```diff
@@ -246,9 +252,176 @@ public SecurityKeyData processSignResponse(SignResponse signResponse, long curre
+        new JsonParser().parse(browserData), "navigator.id.transferAccess", sessionData);
+
 
+    Set<X509Certificate> trustedCertificates = dataStore.getTrustedCertificates();
+    byte[] currentPublicKey = securityKeyData.getPublicKey();
+    X509Certificate currentAttestationCertificate = securityKeyData.getAttestationCertificate();
+    byte[] applicationSha256 = crypto.computeSha256(appId.getBytes());
+
+    checkTransferAccessMessageChain(transferAccessMessages, currentPublicKey,
+        currentAttestationCertificate, trustedCertificates, applicationSha256);
+
+    TransferAccessMessage finalTransferAccessMessage =
+        transferAccessMessages[transferAccessMessages.length - 1];
+    byte[] finalTransferAccessMessageSha256 =
+        crypto.computeSha256(finalTransferAccessMessage.toBytes());
+    byte[] signedBytes =
+        RawMessageCodec.encodeTransferAccessResponseSignedBytes(controlFlags.toByte(), counter,
+            sessionData.getChallenge(), keyHandle, finalTransferAccessMessageSha256);
+
+    currentAttestationCertificate = finalTransferAccessMessage.getNewAttestationCertificate();
+    currentPublicKey = finalTransferAccessMessage.getNewUserPublicKey();
+    
+    Log.info("Verifying signature of bytes " + Hex.encodeHexString(signedBytes));
+    if (!crypto.verifySignature(currentAttestationCertificate, signedBytes, signature)) {
+      throw new U2FException("Signature is invalid for the TransferAccessResponse.");
+    }
+
+    List<Transports> transports = null;
+    try {
+      transports = U2fAttestation.Parse(currentAttestationCertificate).getTransports();
+    } catch (CertificateParsingException e) {
+      Log.warning("Could not parse transports for final TransferAccessMessage " + e.getMessage());
+    }
+
+    SecurityKeyData newSecurityKeyData =
+        new SecurityKeyData(currentTimeInMillis, transports, keyHandle, currentPublicKey,
+            currentAttestationCertificate, /* initial counter value */ counter);
+
+    /*
+     * TODO(alextaka): Make this a transaction:
+     * We remove key first because the original phone will have deleted its key anyway.
+     */
+    dataStore.removeSecurityKey(sessionData.getAccountName(), sessionData.getPublicKey());
+    dataStore.addSecurityKeyData(sessionData.getAccountName(), newSecurityKeyData);
+
+    Log.info("<< processTransferAccess");
+    return newSecurityKeyData;
+  }
+
+  private void checkTransferAccessMessageChain(TransferAccessMessage[] transferAccessMessages,
+      byte[] originalPublicKey, X509Certificate originalAttestationCertificate,
+      Set<X509Certificate> trustedCertificates, byte[] applicationSha256) throws U2FException {
+
+    byte[] currentPublicKey = originalPublicKey;
+    X509Certificate currentAttestationCertificate = originalAttestationCertificate;
+
+    Log.info("  -- Parsed transferAccessMessages -- ");
+
+    for (TransferAccessMessage transferAccessMessage : transferAccessMessages) {
+      Log.info("  sequenceNumber:  "
```

Comment:
I check that when I build the Transfer Access Message chain transferAccessMessages[]. The check isn't done here because the sequence numbers were inserted for parsing, which is done in a previous step.

I don't think code to do this explicitly is important because:
1. They are included in the signatures
2. The signatures need to be in the correct order in order to have this function complete successfully.